calyx/src/passes/compile_sync.rs

```diff
@@ -0,0 +1,380 @@
+use std::collections::HashMap;
+use std::rc::Rc;
+
+use crate::ir::traversal::{Action, Named, VisResult, Visitor};
+use crate::ir::RRC;
+use crate::ir::{self, GetAttributes};
+use crate::{build_assignments, guard, structure};
+use linked_hash_map::LinkedHashMap;
+
+#[derive(Default)]
+/// 1. loop through all control statements under "par" block to find # barriers
+///    needed and # members of each barrier
+/// 2. add all cells and groups needed
+/// 3. loop through all control statements, find the statements with @sync
+///    attribute and replace them with
+///     seq {
+///       <stmt>;
+///       incr_barrier_0_*;
+///       write_barrier_0_*;
+///       wait_*;
+///       restore_*;
+///     }
+///    or
+///     seq {
+///       <stmt>;
+///       incr_barrier_*_*;
+///       write_barrier_*_*;
+///       wait_*;
+///       wait_restore_*;
+///     }
+
+pub struct CompileSync;
+
+/// the structure used to store cells and groups shared by one barrier
+type BarrierMap = LinkedHashMap<u64, (Vec<RRC<ir::Cell>>, Vec<RRC<ir::Group>>)>;
```

Comment:
This type is somewhat of a misnomer because it seems that the two vecs will always contain the same number of elements. If that is the case, then we should replace them with tuples or sized array slices: `[Type; N]`  where `N` is the number of elements.